redis/db/redisbp/prometheus.go

```diff
@@ -0,0 +1,128 @@
+package redisbp
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	promNamespace = "redisbp"
+	subsystemPool = "pool"
+
+	nameLabel = "pool"
+)
+
+// exporter provides an interface for Prometheus metrics.
+type exporter struct {
+	client PoolStatser
+	name   string
+
+	poolHitsCounterDesc     *prometheus.Desc
+	poolMissesCounterDesc   *prometheus.Desc
+	poolTimeoutsCounterDesc *prometheus.Desc
+	totalConnectionsDesc    *prometheus.Desc
+	idleConnectionsDesc     *prometheus.Desc
+	staleConnectionsDesc    *prometheus.Desc
+}
+
+func newExporter(client PoolStatser, name string) *exporter {
+	labels := []string{
+		nameLabel,
+	}
+
+	return &exporter{
+		client: client,
+		name:   name,
+
+		// Upstream docs: https://pkg.go.dev/github.com/go-redis/redis/v8/internal/pool#Stats
+
+		// Counters.
+		poolHitsCounterDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(promNamespace, subsystemPool, "hits_total"),
```

Comment:
is there a reason not to combine these as `cache{result="hit|miss"}` so that sums can be used for totaling?

Reply:
I _think_ the reason is that hits+misses may or may not be the total.

Reply:
When dealing with external values like this where you don't have good control of the underlying implementation I prefer to not guess and expose them as separate metrics.

I'd prefer "cache requests" and "cache misses", but we only have hits and misses.